### PR TITLE
Add GPU implementation of NMSv2 op

### PR DESCRIPTION
**Pull request by samikama**
This PR adds a GPU implementation of NMSv2 Op. It also registers a FakeGPU op for CombinedNonMaxSuppression op to workaround issues encountered due to lack of GPU implementation until a proper GPU implementation can be done based on current GPU kernels.

### CODE REVIEW

**Comment by samikama** [comment]
@tfboyd This is the first part of PRs that would improve performance on object detection networks.

**Comment by samikama** [comment]
Test for new op is blocked by the #28744 since GPU tensors are not correctly transferred to host without it.

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 1]
```diff
@@ -778,5 +778,21 @@ REGISTER_KERNEL_BUILDER(
 
 REGISTER_KERNEL_BUILDER(Name("CombinedNonMaxSuppression").Device(DEVICE_CPU),
                         CombinedNonMaxSuppressionOp<CPUDevice>);
+// REGISTER FakeGPU op for CombinedNonMaxSuppression so that Funcdef'ed versions
+// of it can run.
+REGISTER_KERNEL_BUILDER(Name("CombinedNonMaxSuppression")
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: Please split this part out to a separate PR so we can try to get it into 1.14. The rest of this PR will take longer time to review.

Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: Moved to #28808

**Comment by aaroey** [comment]
Hi @chsigg, could you please help to take a look at this PR?
Thanks.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    const int i = i_block_offset + threadIdx.x;
+    for (int j_thread_offset =
+             NMS_BOXES_PER_THREAD * (blockIdx.y * blockDim.y + threadIdx.y);
+         j_thread_offset < nboxes;
+         j_thread_offset += NMS_BOXES_PER_THREAD * blockDim.y * gridDim.y) {
+      // Note : We can do everything using multiplication,
+      // and use fp16 - we are comparing against a low precision
+      // threshold
+      int above_thresh = 0;
+      bool valid = false;
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Please comment why we need this.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+              j_box.x2 = j_box.x1;
+              j_box.x1 = tmp;
+            }
+            if (j_box.y1 > j_box.y2) {
+              float tmp = j_box.y2;
+              j_box.y2 = j_box.y1;
+              j_box.y1 = tmp;
+            }
+          }
+          const float j_area =
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Please add a comment what is being computed at the high level.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+        d_delete_mask[i * mask_ld + j_thread_offset / NMS_BOXES_PER_THREAD] =
+            above_thresh;
+      }
+    }
+    __syncthreads();  // making sure everyone is done reading smem
+  }
+}
+
+template <typename T, typename Index>
+__global__ void IndexSelect(const CudaLaunchConfig config, const T* original,
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: I would probably just pass in virtual_thread_count (call it num_elements or something).

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+template <typename T>
+__global__ void Iota(const CudaLaunchConfig config, const T offset,
+                     T* to_fill) {
+  for (int idx : CudaGridRangeX(config.virtual_thread_count)) {
+    to_fill[idx] = static_cast<T>(idx) + offset;
+  }
+}
+
+tensorflow::Status nms_gpu(const float* d_desc_sorted_boxes_float_ptr,
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Please use camel case, starting with upper case.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+                           const int N, const float thresh,
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Use lower case please.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+                           int* d_keep_sorted_list, int* h_nkeep,
+                           int* dev_delete_mask, int* host_delete_mask,
+                           OpKernelContext* context, bool flip_boxes) {
+  // d_desc_sorted_boxes_float_ptr is a pointer
+  //    to device memory float array containing the box corners for N boxes.
+  // N is number of boxes.
+  // threshold is the iou threshold for elimination
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Spell out 'intersection-over-union', people might not know.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+  block.x = NMS_BLOCK_DIM;
+  block.y = NMS_BLOCK_DIM;
+  NMSKernel<<<grid, block, 0, device.stream()>>>(
+      d_desc_sorted_boxes, N, thresh, mask_ld, dev_delete_mask, flip_boxes);
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, N);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Use stream_executor::Event

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+  *h_nkeep = nkeep;
+  return Status::OK();
+}
+
+class NonMaxSuppressionV2GPUOp : public OpKernel {
+ public:
+  explicit NonMaxSuppressionV2GPUOp(OpKernelConstruction* context)
+      : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Add comment what the implementation does. Inline (like, above sections below) is also fine.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+      return;
+    }
+    if (num_boxes == 0) {
+      Tensor* output_indices = nullptr;
+      OP_REQUIRES_OK(context, context->allocate_output(0, TensorShape({0}),
+                                                       &output_indices));
+      return;
+    }
+    const int output_size = max_output_size.scalar<int>()();
+    // const float score_threshold_val =
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Remove?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+  auto stream_exec = context->op_device_context()->stream();
+  auto device = context->eigen_gpu_device();
+  dim3 grid, block;
+  int block_size = (N + NMS_BLOCK_DIM - 1) / NMS_BLOCK_DIM;
+  block_size = std::max(std::min(block_size, NMS_BLOCK_DIM_MAX), 1);
+  grid.x = block_size;
+  grid.y = block_size;
+  block.x = NMS_BLOCK_DIM;
+  block.y = NMS_BLOCK_DIM;
+  NMSKernel<<<grid, block, 0, device.stream()>>>(
```
Comment (chsigg) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Use GpuLaunchKernel.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, num_boxes);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
+  device.memcpyDeviceToHost(&host_delete_mask[0], &dev_delete_mask[0],
+                            nto_copy * mask_ld * sizeof(int));
+  CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+  int offset = 0;
+  std::vector<int> h_keep_sorted_list;
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: s/h_keep_sorted_list/h_selected_indices

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+                                thresh, mask_ld, dev_delete_mask));
+  } else {
+    TF_CHECK_OK(GpuLaunchKernel(NMSKernel<false>, grid, block, 0,
+                                device.stream(), d_desc_sorted_boxes, num_boxes,
+                                thresh, mask_ld, dev_delete_mask));
+  }
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, num_boxes);
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: num_to_copy

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+                                device.stream(), d_desc_sorted_boxes, num_boxes,
+                                thresh, mask_ld, dev_delete_mask));
+  } else {
+    TF_CHECK_OK(GpuLaunchKernel(NMSKernel<false>, grid, block, 0,
+                                device.stream(), d_desc_sorted_boxes, num_boxes,
+                                thresh, mask_ld, dev_delete_mask));
+  }
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Is this true only for particular architecture? Shall we make NMS_CHUNK_SIZE architecture specific? Please comment.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+  cudaEventCreate(&copy_done);
+  device.memcpyDeviceToHost(&host_delete_mask[0], &dev_delete_mask[0],
+                            nto_copy * mask_ld * sizeof(int));
+  CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+  int offset = 0;
+  std::vector<int> h_keep_sorted_list;
+  std::vector<int> rmv(mask_ld, 0);
+  memset(host_delete_mask, num_boxes, sizeof(int));
+  while (offset < num_boxes) {
+    const int ncopied = nto_copy;
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: num_copied

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, num_boxes);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
+  device.memcpyDeviceToHost(&host_delete_mask[0], &dev_delete_mask[0],
+                            nto_copy * mask_ld * sizeof(int));
+  CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+  int offset = 0;
+  std::vector<int> h_keep_sorted_list;
+  std::vector<int> rmv(mask_ld, 0);
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Please use more descriptive name

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+    CUDA_CHECK(cudaEventSynchronize(copy_done));
+    if (nto_copy > 0) CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+    for (int i = offset; i < next_offset; ++i) {
+      int iblock = i / NMS_BOXES_PER_THREAD;
+      int inblock = i % NMS_BOXES_PER_THREAD;
+      if (!(rmv[iblock] & (1 << inblock))) {
+        h_keep_sorted_list.push_back(i);
+        int* p = &host_delete_mask[i * mask_ld];
+        for (int ib = 0; ib < mask_ld; ++ib) {
+          rmv[ib] |= p[ib];
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Add some comment here. E.g. if box i is selected, mask out all the boxes that have iou>threshold.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+  int nto_copy = std::min(NMS_CHUNK_SIZE, num_boxes);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
+  device.memcpyDeviceToHost(&host_delete_mask[0], &dev_delete_mask[0],
+                            nto_copy * mask_ld * sizeof(int));
+  CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+  int offset = 0;
+  std::vector<int> h_keep_sorted_list;
+  std::vector<int> rmv(mask_ld, 0);
+  memset(host_delete_mask, num_boxes, sizeof(int));
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: What is the purpose of this memset?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+                                                   &d_selected_indices));
+    Tensor d_sorted_scores;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_FLOAT,
+                                                   TensorShape({num_boxes}),
+                                                   &d_sorted_scores));
+    Tensor d_sorted_boxes;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_FLOAT,
+                                                   TensorShape({num_boxes, 4}),
+                                                   &d_sorted_boxes));
+    Tensor host_nms_mask, d_nms_mask;
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: These two tensors are used only in NmsGpu, please do the allocation inside that function for better abstraction.

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+#include "third_party/cub/device/device_radix_sort.cuh"
+#include "third_party/cub/device/device_segmented_radix_sort.cuh"
+#include "third_party/cub/device/device_select.cuh"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#define NMS_BLOCK_DIM 16
+#define NMS_BLOCK_DIM_MAX 16
+#define NMS_CHUNK_SIZE 2000
+
+#define CUDA_CHECK(result)                                    \
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: @aaroey and @chsigg  are you sure about port::Status? I haven't seen any precedence of port::Status use in any of the kernels in existing TF. Are there plans to switch from tf::status to port::status? Or did you mean tf::Status?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+    config = GetCudaLaunchConfig(num_outputs, device);
+    IndexSelect<<<config.block_count, config.thread_per_block, 0,
+                  device.stream()>>>(config, sorted_indices,
+                                     d_selected_indices.flat<int>().data(),
+                                     (*output_indices).flat<int>().data());
+    CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  }
+};
+REGISTER_KERNEL_BUILDER(
+    Name("NonMaxSuppressionV2").TypeConstraint<float>("T").Device(DEVICE_GPU),
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: That will take time. I will add v3 and v4 in subsequent PRs. Regardless, NMSKernel and NmsGpu are required for #28754 and need to stay as they are. V2 is the closest user of current kernels thus easier to add.

[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: @aaroey Also all these https://github.com/tensorflow/models/search?p=1&q=non_max_suppression&unscoped_q=non_max_suppression can be switched to use v2 op on GPU with a simple regex replace (and should IMHO).

[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: They actually all use v3 ops: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py#L2582. So I would say this implementation probably won't be useful if it's for v2 only.

[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: But isn't V3 with score_threshold=-inf and pad_to_max_output_size=False == V2? I don't see score_threshold and pad_to_max_output being passed around in most places. Just adding _v2 to most of these should give a significant perf boost. Only the one in post_processing.py is making use of V3, the rest can interchangeably use V2 or V3 since V3 with default last 2 arguments is equivalent to v2.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc (hunk 8 above)
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: @csigg I couldn't be able to find any examples of using stream_executor::Event in a similar fashion elsewhere in kernels. Even then, I don't think it is possible to implement the logic by stream_executor::Event since there is no mechanism equivalent to cudaEventSynchronize() implemented in the framework. I can try to spin on event::poll but that would be quite inefficient and would probably hinder the rest of the framework as well due to acquired locks. I would have preferred to use ThenExecute() chaining these but it would require all NMS ops to be converted to AsyncOps as well as a proper threadpool on event manager. Currently all events are executed on single thread and doing work there would block the event infrastructure. I can spawn the work on cpu device thread pool on the event callback but I am not sure if this level of complexity is justified.
How would you propose I would use stream_executor::Event, it is possible that I am missing something obvious.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc (hunk 14 above)
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: It is more related to PCI-E mem-copy and CPU processing time. So it is somewhat flat nowadays. I wouldn't attack NMS_CHUNK_SIZE first if I am to optimize this code. This is first reference implementation and there are already better kernels. However, considering that a GPU NMS implementation is missing, this would bring significant speed up already. I would prefer to have a working implementation at 90% efficiency now instead of waiting for a 95% efficient version next year. We can improve the kernels at a later time.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 22]
```diff
@@ -0,0 +1,498 @@
...
+          Swap(j_box.y1, j_box.y2);
+        }
+        if (OverThreshold(&i_box, &j_box, shared_i_areas[threadIdx.x],
+                          thresh)) {
+          // we have score[j] <= score[i].
+          above_thresh |= (1U << ib);
+        }
+      }
+      if (valid) {
+        d_delete_mask[i * mask_ld + j_thread_offset / NMS_BOXES_PER_THREAD] =
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: would't creating this 2d array be complicated?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc (hunk 21 above)
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: As we discussed in person, I am keeping this and will implement other ops in a later PR.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 23]
```diff
@@ -31,10 +32,23 @@ limitations under the License.
 #define NMS_BLOCK_DIM_MAX 16
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Use constants instead?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 24]
```diff
@@ -227,91 +208,111 @@ __global__ void Iota(const int num_elements, const T offset, T* to_fill) {
...
+    if (num_to_copy > 0) {
+      device.memcpyDeviceToHost(&h_delete_mask[next_offset * bit_mask_len],
+                                &d_delete_mask[next_offset * bit_mask_len],
+                                num_to_copy * bit_mask_len * sizeof(int));
     }
     // Waiting for previous copy
-    CUDA_CHECK(cudaEventSynchronize(copy_done));
-    if (nto_copy > 0) CUDA_CHECK(cudaEventRecord(copy_done, device.stream()));
+    TF_RETURN_IF_CUDA_ERROR(cudaEventSynchronize(copy_done));
+    if (num_to_copy > 0)
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Add {}

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 25]
```diff
@@ -227,91 +208,111 @@ __global__ void Iota(const int num_elements, const T offset, T* to_fill) {
...
+                              config.thread_per_block, 0, device.stream(),
+                              config.virtual_thread_count,
+                              d_nms_mask.flat<int32>().data()));
 
-  // The next kernel expects squares
+  AllocatorAttributes alloc_attr;
+  alloc_attr.set_on_host(true);
+  alloc_attr.set_gpu_compatible(true);
+  TF_RETURN_IF_ERROR(context->allocate_temp(DataType::DT_INT32,
+                                            TensorShape({max_nms_mask_size}),
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: It look like we only need `num_to_copy*2` elements in this buffer, i.e. a circular array. Please either fix it or add comment for the future reference/optimization.

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks a lot for the fix. Please help to address the remaining comments that are still open.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc (hunk 25 above)
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: I will do that. But very skeptical about any benefits such an optimization would bring. Typical ranges of num_boxes in the wild is around 2k to 6k, that would make that host buffer in the order of 6k x 6k/8=4.5MB. for a chunk size of 2000, circular buffer you suggested is 4kx6k/8=3MB . I believe we can save much more memory through other means, not necessarily in this kernel.

**Comment by samikama** [comment]
@aaroey latest changes should address your comments. Please let me know if I missed any.

File: tensorflow/core/kernels/non_max_suppression_op.h [hunk 26]
```diff
@@ -31,8 +32,25 @@ struct NonMaxSuppression {
                   int max_output_size,
                   typename TTypes<int, 1>::Tensor selected_indices);
 };
-
 }  // namespace functor
+#if GOOGLE_CUDA
+extern const int kNmsBoxesPerTread;
+// d_sorted_boxes_float_ptr is a pointer to device memory float array
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.h]: Please describe what this function does in the high level.

Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.h]: Also please make these argument descriptions a list with proper indention for better readability.

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 27]
```diff
@@ -0,0 +1,487 @@
...
+  if (box.x1 > box.x2) Swap(box.x1, box.x2);
+  if (box.y1 > box.y2) Swap(box.y1, box.y2);
+};
+
+// If flip_box is true boxes may have x1>x2 and or y1>y2. If so change the
+// coordinates such that for all boxes x1<x2 and y1<y2. Else boxes should have
+// x1<x2 and y1<y2.
+
+// Starting from highes scoring box, mark any box which has IoU>threshold with
+// given box each thread processes a kNmsBoxesPerThread boxes per stride, and
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Add '.' before 'each thread'?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 28]
```diff
@@ -0,0 +1,487 @@
...
+__device__ EIGEN_STRONG_INLINE void Flipped<true>(Box& box) {
+  if (box.x1 > box.x2) Swap(box.x1, box.x2);
+  if (box.y1 > box.y2) Swap(box.y1, box.y2);
+};
+
+// If flip_box is true boxes may have x1>x2 and or y1>y2. If so change the
+// coordinates such that for all boxes x1<x2 and y1<y2. Else boxes should have
+// x1<x2 and y1<y2.
+
+// Starting from highes scoring box, mark any box which has IoU>threshold with
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Existing comment is about implementation details. Please add a higher level description like "Compute a bit mask `d_delete_mask` where bit `i*bit_mask_len*32+j` is set iff box i and box j has IoU score larger than specific threshold."

File: tensorflow/core/kernels/non_max_suppression_op_test.cc [hunk 29]
```diff
@@ -355,6 +355,178 @@ TEST_F(NonMaxSuppressionV2OpTest, TestEmptyInput) {
...
+            .Finalize(node_def()));
+    TF_EXPECT_OK(InitOp());
+  }
+};
+
+TEST_F(NonMaxSuppressionV2GPUOpTest, TestSelectFromThreeClusters) {
+  MakeOp();
+  AddInputFromArray<float>(
+      TensorShape({6, 4}),
+      {0, 0,  1, 1,  0, 0.1f,  1, 1.1f,  0, -0.1f, 1, 0.9f,
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op_test.cc]: Add a comment saying that these are box coordinates and per-line comments for the corresponding scores (like below), same below.

File: tensorflow/core/kernels/non_max_suppression_op_test.cc [hunk 30]
```diff
@@ -355,6 +355,178 @@ TEST_F(NonMaxSuppressionV2OpTest, TestEmptyInput) {
   test::ExpectTensorEqual<int>(expected, *GetOutput(0));
 }
 
+#if GOOGLE_CUDA
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op_test.cc]: Existing BUILD target doesn't run gpu tests, please move these test cases to a new `non_max_suppression_op_gpu_test.cc` file and add a tf_cuda_cc_test target for it.

**Review by aaroey — CHANGES_REQUESTED** [review]

**Review by aaroey — APPROVED** [review]
Thanks so much for all the fixes!

@chsigg just checking if you have any other comments.

**Comment by aaroey** [comment]
Sorry for the delay, there are some internal test failures and I'm still trying to fix them.

**Comment by ppwwyyxx** [comment]
I believe this implementation is wrong: it does not agree with the CPU version of NMS Op.

In this implementation, when computing the area in IOU, it uses `(x2-x1+1)*(y2-y1+1)`, as can be seen at: https://github.com/tensorflow/tensorflow/blob/e3062d1924c5d07d07440bc8a7de8d97615a6f52/tensorflow/core/kernels/non_max_suppression_op.cu.cc#L96-L98

However, in the CPU version, it uses `(x2-x1)*(y2-y1)`, as can be seen at: https://github.com/tensorflow/tensorflow/blob/e3062d1924c5d07d07440bc8a7de8d97615a6f52/tensorflow/core/kernels/non_max_suppression_op.cc#L126-L129

For many inputs this may not have an effect at all. But for certain inputs the two versions will produce inconsistent results.

If your goal is to run object detection models, note that the "+1" is a legacy issue and we're trying to avoid the version with "+1" in Facebook. See [this PR](https://github.com/pytorch/pytorch/pull/20550) that handles "+1" in caffe2.

**Comment by aaroey** [comment]
@samikama would you please fix the issue @ppwwyyxx mentioned above?

**Comment by SpaceInvader61** [comment]
@samikama I tried to use your kernel from inside python and I am getting a segmentation fault by running this simple script:

```py
import tensorflow as tf

tf.enable_eager_execution()

boxes = tf.constant([[1.0, 1.0, 1.0, 1.0]], dtype=tf.float32)
scores = tf.constant([1.0], dtype=tf.float32)
max_output_size = tf.constant(10, dtype=tf.int32)
iou_threshold = tf.constant(0.7, dtype=tf.float32)
score_threshold = tf.constant(float('-inf'), dtype=tf.float32)

from tensorflow.python.ops import gen_image_ops

with tf.device("/device:GPU:0"):
    print("Start")
    x = gen_image_ops.non_max_suppression_v2(boxes, scores, max_output_size, iou_threshold, score_threshold)
    print("End")
    print(x)
```

```bash
docker run --runtime=nvidia -it -v $PWD:/tf -w /tf tensorflow/tensorflow:nightly-gpu-py3
 python pyscript.py
```

The output is
```
Start
Segmentation fault
```

Correct me if I am doing smth wrong

**Comment by samikama** [comment]
@ppwwyyxx Thanks for catching that. I made the fixes to support both legacy case and CPU identical implementation in #30893.
@SpaceInvader61 It looks like we missed that some input tensors need to be host tensors when they are changed from attributes to tensors. Can you try with #30893? Also your example is probably not really making use of the nms_v2 since the signature of nms_v2 is 
> non_max_suppression_v2(boxes, scores, max_output_size, iou_threshold, name=None)
Another point is you are passing a box with 0 surface area and that is the only box. Even though there is a single box test in the test suite, we didn't have an invalid box test. I will add the fix for it in an upcoming PR.

**Comment by SpaceInvader61** [comment]
@samikama with .HostMemory() it works, thank you so much! :+1:

**Comment by AkshayKhatriKodiak** [comment]
Has this made it into any of the tensorflow releases? As far as I know, it wasn't included in 1.13 and 1.14. How about in tensorflow 2.0?
